Fix relatable scope used for determining possible work packages to create a relation to.

ulferts opened the pull request:
Both allows and prevents relations to be constructed in additional structures, e.g.:

* [For a work package with a predecessor. With that predecessor having a parent: That parent can now be set as the parent of the original work package.](https://community.openproject.org/wp/44827) 
* For a work package with a predecessor. With that predecessor having a predecessor of its own: That predecessor's predecessor can no longer be set as a parent.
* For a work package with a child. With that child having a successor: That successor can no longer be set to be the parent.
* For a work package with a parent currently. With that parent having a predecessor: The predecessor can now be set to be the new parent.  
 
## TODO

* [x] additional documentation
* [x] additional testing

ulferts committed:
fix relatable scope

ulferts commented on app/models/work_packages/scopes/relatable.rb:
```diff
@@ -88,39 +124,41 @@
-        else
-          scope
-        end
+        where("work_packages.id NOT IN (#{Arel.sql(sql)})")
```
All the inputs are escaped before making it into the SQL.

cbliard commented on app/models/work_packages/scopes/relatable.rb:
```diff
@@ -29,9 +29,9 @@ module Relatable
 
     class_methods do
-      # Returns all work packages that are relatable to the provided work package with a relation of the provided type.
+      # Returns all work packages that are relatable to the provided work package via new a relation of the provided type.
```
I'm not sure to understand this sentence.

I'd prefer this (if I correctly understood):

Returns all work packages for which a new relation of the provided type can be created.

ulferts replied:
> Returns all work packages for which a new relation of the provided type can be created.

Can we maybe go with 

"Returns all work packages which can become part of a new relation from or to the provided work package where the new relation would receive the provided type (e.g. 'blocks')."

cbliard replied:
not sure about the "from or to" as I had more the impression that it was one or another (either precedes relation or follows relation for instance), but that definitely reads better than before.

ulferts replied:
You are right in so far as that only one direction is every actually searched for when executing the code. However the direction depends on the relation type provided so it is capable of handling both directions.

ulferts replied:
Done in a29cf73bb2ea3ba08594b53051c37312639b1465

cbliard commented on app/models/work_packages/scopes/relatable.rb:
```diff
@@ -40,33 +40,114 @@ module Relatable
       # * Ancestor/descendant: relations cannot be drawn between ancestor and descendants. It is important to note though,
       #   that relations can be created between any two work packages within the same tree as long as they are not in a direct
       #   or transitive ancestor/descendant relationship. So a relation between siblings is just as possible as one between
-      #   aunt and nephew.
+      #   aunt and nephew. Additionally, a transitive relationship which leaves the tree for any one hop, is exempt from that rule
+      #   so it is possible to have child --- blocks ---> other --- blocks ---> parent.
```
In this sentence: "It is important to note though, that relations can be created between any two work packages within the same tree as long as they are not in a direct **or transitive** ancestor/descendant relationship. So a relation between siblings is just as possible as one between and nephew."

I'm not sure to understand why a relation cannot be drawn between two work packages in a transitive ancestor/descendant relationship. And I'm not sure what a transitive ancestor/descendant relationship is exactly.

And it can be fooled:
In this configuration
```
parent
+ child
  + child child
relates-to-parent (with a relates-to relation with parent)
```
![image](https://user-images.githubusercontent.com/176055/203503017-26bce7fa-447e-46ff-a084-d77b7897b57d.png)

It is not possible to set 47-parent as a parent of 53-relates-to-parent, because they already have a direct ancestor/descendant relationship. This is ok and conforms to the doc.

But it's possible to set 48 as a parent, despite having a transitive ancestor/descendant relationship (if I understood what transitive means correctly). Is it intended or is it a bug?

![image](https://user-images.githubusercontent.com/176055/203503547-ff9d91bc-eec8-4224-950c-4243a2896e27.png)

ulferts replied:
> And I'm not sure what a transitive ancestor/descendant relationship is exactly.

My understanding of a transitive relationship as opposed to a direct relationship is that a transitive relationship consists of 2 or more direct relationships. So the work packages involved might not be directly connected but are in effect still connected. In the example above, 47-"parent" would be transitively related to 50-"child child" since they are both connected to 48-"child". For ancestors/descendants that distinction admittedly might not be as important as it is for other relations like follows/precedes.

> But it's possible to set 48 as a parent, despite having a transitive ancestor/descendant relationship (if I understood what transitive means correctly). Is it intended or is it a bug?

This is a valid relation since for the 'relates' type, some limitations do not apply (As you have noted yourself in a different comment)

cbliard replied:
Then this whole paragraph is wrong:

> * Ancestor/descendant: relations cannot be drawn between ancestor and descendants. It is important to note though, that relations can be created between any two work packages within the same tree as long as they are not in a direct or transitive ancestor/descendant relationship. So a relation between siblings is just as possible as one between aunt and nephew. Additionally, a transitive relationship which leaves the tree for any one hop, is exempt from that rule so it is possible to have child --- blocks ---> other --- blocks ---> parent.

The first sentence: "relations cannot be drawn between ancestor and descendants.", and yet a "relates-to" is ok... This should be mentioned here.

It could even be shorter: 

* Ancestor/descendant: relations can be created between any two work packages within the same tree as long as they are not in a ancestor/descendant relationship.
  For instance, a relation between siblings is possible, a relation between aunt and nephew is possible, whereas a relation between a parent and a child is not possible, and a relation between grand parent and grand child is not possible either.
  The only exception is for the "relates" relation which can apply to ancestors / descendants, but can not apply to direct parent / child because of the previous rule.

ulferts replied:
I read up on your suggestion a bit late so I already changed the paragraph recently and did not include your suggestion. Can you please take a look to see if that new text is better already.

The exception is still not excluded since there is an extra paragraph just for the exceptions and additions to the rules which I think makes it easier to understand the rules first, and then see to the special cases. Especially if the additional rules for the PARENT relationship were to be included as well, it would be way harder to understand.

cbliard commented on app/models/work_packages/scopes/relatable.rb:
```diff
@@ -40,33 +40,114 @@ module Relatable
       #   This means that starting from the work package both the descendants as well as the ancestors are considered. That
       #   way, relations like
-      #     Child 1 --- parent ---> Parent 1 ---> follows ---> Parent 2 ---> child ---> Child 2 --- follows ---> Child 1
+      #     Child tree 1
+      #       --- parent ---> Parent tree 1
+      #       --- follows ---> Parent tree 2
+      #       --- child ---> Child tree 2
+      #       --- follows ---> Child tree 1
```
I do not understand this schema. How can a tree be a parent of another tree?
Is it this?
```
Parent --> Parent
   |         |
   v         v
 Child --> Child

--> is a follows/precedes relation
```

ulferts replied:
Following your notation, it would be:

```
Parent --> Parent
   |         |
   v         v
 Child <-- Child

--> is a follows/precedes relation
```

cbliard commented on app/models/work_packages/scopes/relatable.rb:
```diff
@@ -40,33 +40,114 @@ module Relatable
       # There are a couple of exceptions and additions to the limitations outlined above for the following types:
-      # * Relation::TYPE_RELATES: Since this is essentially unrelated and does not carry a lot of semantic, the work packages
+      # * Relation::TYPE_RELATES: Since this is essentially undirected and does not carry a lot of semantic, the work packages
       #   are simply somehow related, such relations do not follow the "non circular" nor the "ancestor/descendant" rule.
```
ah, that's probably why the relates-to relation I described above was possible between grand-parent and grand-child. But then why is it not possible between parent-child? Is it because a relation already exists?

It's not limited to relates-to relation. It's possible to make a blocks or requires relation between a grand parent and grand child by creating the relation first, then drag and dropping under child (or setting parent).

ulferts replied:
Right, for relates, being able to create such a relation is ok. If it is possible for blocks, then that it is a bug.

cbliard commented on app/models/work_packages/scopes/relatable.rb:
```diff
@@ -40,33 +40,114 @@ module Relatable
+      #   predecessors/successors can be related to. But they still need to be considered in the code since they might be
+      #   part of an almost completed circle which would be closed if a work package is added as a parent. E.g. in
+      #     WP1 --- follows ---> WP2 --- parent ---> WP3 --- follows ---> WP4
+      #   WP4 is not a valid parent candidate whereas WP3 is.
```
A valid parent candidate for which work package? WP1?
But if WP3 is a child of WP2, I do not understand why WP1 can be a child of WP3. That should not be possible. What am I getting wrong here?

ulferts replied:
Yes, the queried for work package would be WP1. I should have made that clearer in the text.

There is a flaw in my example here in that the parent relation between WP2 and WP3 means to indicate that WP3 is a parent of WP2 and not the other way around. That is why WP1 can become a child of WP3 which would turn its predecessor WP2 into a sibling. But WP4 cannot be a parent since that would result in 

```
WP3 --> WP4
 |       |
 v       v
WP2 <-- WP1

```

I suggest to fix the flaw and also to show the resulting trees which might make it more understandable.

ulferts replied:
Done in a29cf73bb2ea3ba08594b53051c37312639b1465

cbliard commented on app/models/work_packages/scopes/relatable.rb:
```diff
@@ -40,33 +40,114 @@ module Relatable
+      #       <-- follows --- Successor WP
+      #       --- parent ---> Successor's parent WP
+      #       --- follows --> Successor's parent's predecessor WP
+      # all work package are potential successors.
```
```suggestion
      # all work packages are potential successors.
```

cbliard commented on app/models/work_packages/scopes/relatable.rb:
```diff
@@ -40,33 +40,114 @@ module Relatable
-      # exist.
+      # For FOLLOWS/PRECEDES relationships both directions need to be taken into account, and again, this includes those of the
+      # hierarchy. This prevents creating relations in a structure like this:
+      #     queried WP
+      #       --- follows ---> Predecessor WP
+      #       --- parent ---> Predecessor's parent WP
+      #       <-- follows --- Predecessor's parent's successor WP
```
I have a hard time reading this kind of schema.
Should it be expanded like this?
```
      #      queried WP --- follows ---> Predecessor WP
      #      queried WP --- parent ---> Predecessor's parent WP
      #      queried WP <-- follows --- Predecessor's parent's successor WP
```

Maybe it's hard for me because when reading `A --- follows ---> B`, here A is on the left and B on the right, but on a timeline B would be on the left and A on the right.

ulferts replied:
I can try to pick up your notation and see if that is easier for the two of us to understand. But I fear that the direction change will be inevitable at least for some of the invalid trees.

ulferts replied:
Updated the docs in a29cf73bb2ea3ba08594b53051c37312639b1465

cbliard commented on app/models/work_packages/scopes/relatable.rb:
```diff
@@ -40,33 +40,114 @@ module Relatable
       def relatable(work_package, relation_type)
         return all if work_package.new_record?
 
         scope = not_having_directed_relation(work_package, relation_type)
-                  .not_having_direct_relation(work_package)
```
The names `not_having_directed_relation` and `not_having_direct_relation` are close. Are they in the same order as in the method documentation?

```
      # The implementation focuses on excluding candidates. It does so in two parts:
      #   * Excluding all work packages with which a direct relation already exist.
      #   * Excluding work packages that are related transitively.
```
It looks like they're not
And the order of usage is not the same as the order of definition below.

I like when order is the same everywhere: doc, invocation, definition. it makes it easier to understand. Can you swap `not_having_directed_relation` and `not_having_direct_relation` calls so that everything is in the same order?

And maybe renaming `not_having_directed_relation` as `not_having_transitive_relation` would better relate to the method description above.

ulferts replied:
I restructured the code quite a bit in a7c3914b77a7b43c9a0bf5c19df2c10969889db1. The order in the code is now the same as it is in the documentation. The names changed significantly since the code now differentiates directly between PARENT relations and the rest.

cbliard commented on spec/models/work_packages/scopes/relatable_spec.rb:
```diff
@@ -27,6 +27,8 @@
 describe WorkPackages::Scopes::Relatable, '.relatable scope' do
+  create_shared_association_defaults_for_work_package_factory
+
   let(:project) { create(:project) }
```
This change:
```suggestion
  let(:project) { project_with_types }
```

makes the spec time go from 16 seconds to 8.5 seconds as it reuses the project created in a `shared_let` when calling `create_shared_association_defaults_for_work_package_factory`.

ulferts replied:
Makes total sense. But I'll just try to remove all the references to `project` since that should be handled by the `create_shared...` unless I am mistaken and will trim down the spec.

cbliard replied:
yes, that's even better

cbliard commented on app/models/work_packages/scopes/relatable.rb:
```diff
@@ -40,33 +40,114 @@ module Relatable
-                  .not_having_direct_relation(work_package)
-                  .where.not(id: work_package)
+                  .not_having_direct_relation(work_package, relation_type)
+                  .where.not(id: work_package.id)
```
I would almost define a `not_having` method and use it instead of the `where.not`

```
def not_having(work_package_id)
  where.not(id: work_package_id)
end
```

This way all calls have the same "shape" and the same abstraction level.

cbliard commented on app/models/work_packages/scopes/relatable.rb:
```diff
@@ -40,33 +40,114 @@ module Relatable
+                  .not_having_direct_relation(work_package, relation_type)
+                  .where.not(id: work_package.id)
+
+        # On a parent relationship, explicitly remove the former parent (which might be the current one as well)
+        # from the list of work packages one can relate to. This is not strictly necessary since it would not
+        # cause faulty relationships but doing it removes the parent from places where it should not show up,
+        # e.g. in an auto completer.
+        if relation_type == Relation::TYPE_PARENT && work_package.parent_id_was
+          scope = scope.where.not(id: work_package.parent_id_was)
+        end
```
Why not move this into `not_having_direct_relation`?

ulferts replied:
I did this now though the implementation has changed quite a bit in a7c3914b77a7b43c9a0bf5c19df2c10969889db1. I am not sure if it fits in the newly created `not_having_potential_tree_relation` just on the basis of the name but it makes the code easier to read.

cbliard reviewed (commented):
It seems to work but here are some edge cases for which I'm not sure if that's wanted or not. Please see comments.

I have troubles fully understanding how it works, and I did not get into the core of the CTE yet. I'll get into it once you've answered my questions (in comments) regarding how it works.

Another thing I do not like is error messages, but this can probably be fixed in another PR. Here is the issue:

When in this configuration:
```
parent
+ child
related-to-parent (with a 'relates-to' relation with parent)
```
![image](https://user-images.githubusercontent.com/176055/203501022-eb34a22d-3ce4-479a-afda-483bb3bc07c2.png)

If trying to drag and drop the `related-to-parent` work package under parent, then it fails with the error `Parent A work package cannot be linked to one of its subtasks.`. This gives no clue about why it does not work and what should be done to make it work. Here the 'relates-to' relation should be removed before making it a child of 'parent'.

ulferts committed:
remove superfluous explicit references to project in spec

ulferts committed:
lifting precedes/follows constraint and updating documentation

ulferts committed:
prevent parent relating to a child of a related to work package

cbliard commented:
Bug found: **With two parent/child trees, it's impossible to have parent following each other, and children following each other too**

It should be possible to create this structure:
```
      #                 WP parent1 <────follows──── WP parent2
      #                    │                            │
      #                    │                            │
      #                 hierarchy                    hierarchy
      #                    │                            │
      #                    v                            v
      #                  WP child1 <────follows──── WP child2
```

I could create this
```
      #                 WP parent1                   WP parent2
      #                    │                            │
      #                    │                            │
      #                 hierarchy                    hierarchy
      #                    │                            │
      #                    v                            v
      #                  WP child1 <────follows──── WP child2
```
But then from the GUI I could not create the "follows" relation between the parents.

cbliard commented:
Maybe bug found: **Impossible to create relates_to relation between ancestors/descendants**

Doc says:
> There are a couple of exceptions and additions to the limitations outlined above for the following types:
> * Relation::TYPE_RELATES: Since this is essentially undirected and does not carry a lot of semantic, the work packages
>   are simply somehow related, such relations do not follow the "non circular" nor the "ancestor/descendant" rule.

But with tthis hierarchy:
```
main
  ┗━ child
       ┗━ grandchild
```
It's not possible to create a "relates_to" relation between main and grandchild.

(note than with 3 work packages following each other, it's possible to create a relates_to from the first to the last. This part is working as expected)

cbliard commented:
Strange behavior:

When setting 3 work packages following each other like this:
```
WP1 <────follows────  WP2 <────follows──── WP3
```

In GUI in the relations tab:
* when selecting WP1 and trying to create a follows relation, WP2 and WP3 are not listed. That's ok: that would create a circular dependency.
* when selecting WP3 and trying to create a precedes relation, WP2 is not listed, but WP1 is listed. That's weird: that would create a circular dependency.
  * when clicking WP1 to create the circular dependency, an error message is displayed: "This relation would create a circular dependency.". So it's prevented but it's strange that the behavior is not symmetric here.

(there is the same kind of issue with children: I'll describe it in another comment)

cbliard commented:
Maybe bug found: **a work package can't follow a parent and its child.**

In this configuration
```
    WP1 <────follows──── WP3
     │
     │
  hierarchy
     │
     v
    WP2
```
I thought it would be possible for WP3 to follow WP2, but the GUI does not allow me to do so.
WP3 can follow WP1, or WP2, but not both.

Is it expected? It does not create any circular dependency.

cbliard commented:
Strange behavior:

A bit similar to the one with 3 work packages following each other

In this configuration:
```
    WP1 <────follows──── WP3
     │
     │
  hierarchy
     │
     v
    WP2
```

* When selecting WP3, and clicking "Set parent", neither WP1 or WP2 is proposed as it would create a circular dependency.
* When selecting WP2, and clicking "Add existing child", WP1 is not proposed but WP3 is available, even if it would create a circular dependency.
  * If clicking to create the child relation, it fails with the following error: "Parent A work package cannot be linked to one of its subtasks.". So it's fool proof but here too it's strange to not have behavior symmetry.

cbliard commented:
@ulferts I'm done with testing it. Codewise, I do not have any additional remarks. You have greatly improved the documentation of the behavior. I'll assign you back in the work package.

ulferts committed:
remove canoncial cast in relatable filter

The `relatable` scope can handle all relation types. Casting it to the canonical type yields the wrong result for the non canonical types, e.g. precedes, since it in effect would search for follows

ulferts committed:
add test case to relatable scope

ulferts committed:
only have single rule - direct relation - apply to relates

ulferts committed:
include child relation type

ulferts committed:
remove unused column

ulferts committed:
support the child operator properly

ulferts commented:
@cbliard, I believe to have addressed every one of your remarks. 

There were three flaws in my previous code:

* relatable indeed had too many rules applied. Now it just short circuits and does not do any CTE calculation any more.
* The notion of `child` was not present in the relatable scope which works as the inverted counterpart to `parent`. Because of that, the relatable scope was not capable to correctly calculate candidates to become children. Gladly, this really is just the inverted version to `parent`.
* The `RelatableFilter` used to always call the `relatable` scope with the canonical relation type which lead to the directing getting lost when e.g. attempting to create a 'precedes' relation.  

Can you please check the PR again?

cbliard commented:
@ulferts  All the previous cases are fixed :+1: 
Special thanks to have made different commits for each bug fix. It makes reviewing easier.

I found another bug:

In this configuration
```
                         WP2 <────follows──── WP1
                          │
                          │
                       hierarchy
                          │
                          v
   WP4 <────follows────  WP3
    │       
    │       
 hierarchy  
    │       
    v       
   WP5      
```

When WP1 is selected, setting parent to WP5 is possible, and it should not be (for the other wps it's ok they're not selectable). There is no guard this time, so it creates a circular relation and it produces the error "Stack level too deep" in the `SetScheduleService` when computing dependencies.

Strangely enough, the reverse is not possible: when WP5 is selected, it's not possible to set any other work package in the tree as existing child.

ulferts committed:
fix includes_hierarchy condition

ulferts commented:
@cbliard that scenario should be fixed now. I apparently made a simplification before that I shouldn't have.

cbliard commented:
@ulferts previous one is fixed.

But there is another one in the same configuration:
```
                         WP2 <────follows──── WP1 <────follows──── WP0
                          │
                          │
                       hierarchy
                          │
                          v
   WP4 <────follows────  WP3
    │       
    │       
 hierarchy  
    │       
    v       
   WP5      
```

When WP1 is selected (or WP0), adding existing child WP4 is possible, and it should not be. There is something guarding from doing it and the error message `Parent A work package cannot be linked to one of its subtasks. ` is displayed after trying.

Here again, trying to add the same relation from WP4 (setting the parent to WP1) will not be possible.

ulferts commented:
@cbliard and I agreed to work on the latest bug separately and merge this PR now, since what the PR offers is better than the checks in place before.
